### DISCUSSION TITLE
Debugging the rainbow screen

### DISCUSSION THREAD

**Discussion by SteveSpencer** [rsta2/circle#484] [Q&A] [answered]
I have written/ported a driver for the RoboPeak USB DIsplay (no longer available!) and am testing it on a Pi Zero, with serial logging.
The test application I have written works fine (so far), and I have put hooks into the device to allow it to be used with LVGL, since that seemed a useful thing for me to do.

So, I set up a uSD card, copied my simple test app, which just paints some rectangles and handles touch events, and that works quite nicely. I proceeded to create a copy of the lvgl class, and modify it to work with my driver. I have a sample app that uses my modified class in place of the CLVGL one, and built it.

I copy the resultant kernel.img to the SD card, and i get the rainbow screen on the HDMI display, but my USB screen isn't initialising. I know this, because on power up it displays 'waiting for signal' until I actually send it a command. 

How early in main/kernel can I put logging to see what is happening?

I am clearly doing something wrong, as the original LVGL sample works fine, so there must be something in my version of the LVGL constructor that is causing an issue.

**Comment by rsta2** [comment, marked as answer]
The following classes must be instantiated, before logging can be used:

* CDeviceNameService
* CSerialDevice (target for the logger)
* CLogger

You must call `Initialize()` for the following classes, before logging can be used:

* CSerialDevice (target for the logger)
* CLogger

`CLogger` can be constructed without a reference for a `CTimer` object. It will not show the time then.

In your case I would instantiate the class `CLVGL` not as a member of `CKernel`, but as a local variable in `CKernel::Run()`. Then you can do all necessary initialization for the logger before.

**Reply by SteveSpencer** [reply]
Yes, I did that, and discovered my error. I had tried to instantiate another instance of CDeviceNameService in order for my LVGL class to locate the display and touch devices, instead of using CDeviceNameService::Get() to get the single instance. As this was a class member in the LVGL class (another rookie error), on construction it asserts, but the logging isn't yet initialized (as per your reply), so the code neatly fails, but quietly...

Thanks.
